### PR TITLE
chore(curriculum): add spam filter workshop to new cert

### LINKED ISSUE #624: chore: add spam filter workshop to main repo
We are using the spam filter project from the JS beta version.

We need to add it to the main repo with steps and tests.

Take a look at the original here 
https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/learn-regular-expressions-by-building-a-spam-filter/step-36

See if there are areas were we can update the tests to use less regex testing and have campers build our more of the functions on their own and accept multiple answers

### PR DESCRIPTION
**Pull request by Dario-DC**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes https://github.com/freeCodeCamp/CurriculumExpansion/issues/624

### CODE REVIEW

**Review by jdwilkin4 — COMMENTED** [review]
For the workshops, we don't need to be as strict. 

So for example, where it says `You should use const` we should allow any valid answer like let, const or even var.
The only time we should care is if they need to reassign a value, then we should test for let.

As for accessing elements for the DOM, they should be able to use `getElementById` or `querySelector`. We shouldn't need to specifically test for `getElementById`. We should only test for the final result. How they go about it is up to them.

I realize that you might have used the same tests and steps from the best JS version, but there are a lot of issues with the beta JS version. So it is best to fix those now 👍 

Same issue for these click events, we should just test the final result instead of breaking it down so much into small steps and using a bunch of regex testing.

Overall, your approach to these workshops should be more similar to how you would approach the labs. In the labs, you are testing for final results and very rarely using regex testing unless it was needed. This allows for more valid answers and less issues with tests.

**Comment by Dario-DC** [comment]
This is exactly the project from the beta version. I haven't changed anything yet :+1: @jdwilkin4

File: curriculum/challenges/english/25-front-end-development/workshop-spam-filter/641cf198ec366c33d6504854.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+The <dfn>alternate sequence</dfn> `|` can be used to match either the text on the left or the text on the right of the `|`. For example, the regular expression `/yes|no/` will match either `yes` or `no`.
+
+Update your `helpRegex` to match either `please help` or `assist me`.
+
+# --hints--
+
+Your `helpRegex` should use the `|` alternate sequence.
+
+```js
+assert.includes(helpRegex.toString(), "|");
```
Comment (Dario-DC) [inline curriculum/challenges/english/25-front-end-development/workshop-spam-filter/641cf198ec366c33d6504854.md:19]:
```suggestion
assert.include(helpRegex.toString(), "|");
```

**Comment by ilenia-magoni** [comment]
reminder: use `[skip ci]` in the commit message when you don't want the tests to run yet

File: curriculum/challenges/english/25-front-end-development/workshop-spam-filter/641cd91d28bebe226f765d86.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  messageInput.value = "";
+  checkMessageButton.dispatchEvent(new Event("click"));
+  assert.lengthOf(testArr, 1);
+  assert.strictEqual(testArr[0], "Please enter a message.")
+} finally {
+  alert = temp;
+}
+```
+
+When you click your `checkMessageButton` you should not alert the user if `messageInput` is not empty.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/workshop-spam-filter/641cd91d28bebe226f765d86.md:35]:
```suggestion
When you click your `checkMessageButton` you should only alert the user if `messageInput` is empty.
```

it works either way but double negation is confusing

File: curriculum/challenges/english/25-front-end-development/workshop-spam-filter/64220fb017c57d20612de8b8.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+```js
+assert.match(dollarRegex.source, /\(hundred\|thousand\|million\|billion\)\?/);
+```
+
+Your `dollarRegex` should match `1000 dollars`.
+
+```js
+const dollarStr = '1000 dollars'
+assert.deepEqual(dollarStr.match(dollarRegex), dollarStr.match(/[0-9]+ (hundred|thousand|million|billion)? dollars/i));
+```
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/workshop-spam-filter/64220fb017c57d20612de8b8.md:33]: this doesn't match, you need two spaces in the string to be matched by this regex.

**Review by majestic-owl448 — COMMENTED** [review]
I have reviewed up to step 16.

See the image here for the issue in step 16
image

File: curriculum/challenges/english/25-front-end-development/workshop-spam-filter/64220fb017c57d20612de8b8.md (hunk 3 above)
Comment (Dario-DC) [reply curriculum/challenges/english/25-front-end-development/workshop-spam-filter/64220fb017c57d20612de8b8.md:33]: I'm just going to remove this test. In the following step, it's mentioned that this won't match `10 dollars`. So it shouldn't be an issue.

Comment (ilenia-magoni) [reply curriculum/challenges/english/25-front-end-development/workshop-spam-filter/64220fb017c57d20612de8b8.md:33]: Seems reasonable!

File: curriculum/challenges/english/25-front-end-development/workshop-spam-filter/641ceed81533263283835c3d.md [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+title: Step 8
+challengeType: 0
+dashedName: step-8
+---
+
+# --description--
+
+Instead of using the `.match()` method, you can use the `.test()` method of a regular expression to test if a string matches the pattern. Unlike `.match()`, `.test()` returns a boolean value indicating whether or not the string matches the pattern.
+
+Update your `isSpam()` function to use the `.test()` method of `helpRegex` to test if `msg` is a match.
```
Comment (zairahira) [inline curriculum/challenges/english/25-front-end-development/workshop-spam-filter/641ceed81533263283835c3d.md:12]:
```suggestion
Update your `isSpam()` function to use the `.test()` method on `helpRegex` to test if `msg` is a match.
```

**Review by zairahira — COMMENTED** [review]
Reviewed till step 10.

File: curriculum/challenges/english/25-front-end-development/workshop-spam-filter/6421f98f4999d1179ce37cb4.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+Arrays have a `.some()` method. Like the `.filter()` method, `.some()` accepts a callback function which should take an element of the array as the argument. The `.some()` method will return `true` if the callback function returns `true` for at least one element in the array.
+
+Here is an example of a `.some()` method call to check if any element in the array is an uppercase letter.
+
+```js
+const arr = ["A", "b", "C"];
+arr.some(letter => letter === letter.toUpperCase());
+```
+
+Use the `.some()` method to check if testing your `msg` on any of your `denyList` regular expressions returns `true`.
```
Comment (zairahira) [inline curriculum/challenges/english/25-front-end-development/workshop-spam-filter/6421f98f4999d1179ce37cb4.md:19]:
```suggestion
Update the `isSpam()` function to use the `some()` method, which will check if testing `msg` against any of the regular expressions in `denyList` returns `true`." 
```

File: curriculum/challenges/english/25-front-end-development/workshop-spam-filter/6421f98f4999d1179ce37cb4.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
+---
+id: 6421f98f4999d1179ce37cb4
+title: Step 11
+challengeType: 0
+dashedName: step-11
+---
+
+# --description--
+
+Arrays have a `.some()` method. Like the `.filter()` method, `.some()` accepts a callback function which should take an element of the array as the argument. The `.some()` method will return `true` if the callback function returns `true` for at least one element in the array.
```
Comment (zairahira) [inline curriculum/challenges/english/25-front-end-development/workshop-spam-filter/6421f98f4999d1179ce37cb4.md:10]:
```suggestion
Arrays have a `some()` method. Like the `filter()` method, `some()` accepts a callback function which should take an element of the array as the argument. The `some()` method will return `true` if the callback function returns `true` for at least one element in the array.
```

**Review by zairahira — COMMENTED** [review]
Reviewed all the steps.

**Comment by moT01** [comment]
A PR has been merged that renamed the certification from "Front End" to "Full Stack". This PR may need an update and a few changes made: [see comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/56694#issuecomment-2444473378)

**Comment by moT01** [comment]
A PR has been merged that moves the block order from the `meta.json` file to the `full-stack.json` file.](https://github.com/freeCodeCamp/freeCodeCamp/pull/56795) This PR will need an update, and the order will need to be removed from the `meta.json` file and moved to the new file.

**Review by zairahira — APPROVED** [review]

**Review by moT01 — APPROVED** [review]

**Pull request merged**